{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.metrics import make_scorer, accuracy_score, f1_score, roc_auc_score, precision_score, recall_score, classification_report\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import argparse\n",
    "from sklearn.model_selection import train_test_split\n",
    "from catboost import CatBoostClassifier\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.feature_selection import mutual_info_classif\n",
    "from sklearn.model_selection import cross_validate\n",
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "from datetime import datetime, timedelta\n",
    "# from xgboost import XGBClassifier\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "# pd.options.display.max_columns = None\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PREPROCESS DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fdir_raw = Path(\"/home/ar3/Documents/PYTHON/RNASeqAnalysis/data/raw/\")\n",
    "data = pd.read_csv(fdir_raw/'Geuvadis.all.csv', index_col=0)\n",
    "# data.rename(columns={\"Unnamed: 0\": \"trascripts\"})\n",
    "data = data.T\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_header = pd.read_csv(fdir_raw/'Geuvadis.SraRunTable.txt',index_col=0)\n",
    "data_header = data_header[['Sex', 'Experimental_Factor:_population (exp)']]\n",
    "# # data_header.set_index('Run', inplace=True)\n",
    "data_header.head()\n",
    "\n",
    "# data_header.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_by_non_zero_median(df):\n",
    "    print(df.shape)\n",
    "    df_median = df.median()\n",
    "    if (df_median == 0).any():\n",
    "        cols_to_drop = df.columns[df_median == 0]\n",
    "        print(len(cols_to_drop),\n",
    "              \" features will be removed, due to a zero median value\")\n",
    "        df = df.drop(columns=cols_to_drop)\n",
    "        print(\"Current dataset size: \", df.shape)\n",
    "        return df\n",
    "\n",
    "    print(\"Zero median columns aren't found\")\n",
    "    print('Dataset shape: ', df.shape)\n",
    "    return df\n",
    "\n",
    "data = filter_by_non_zero_median(data).astype('float32')\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import pointbiserialr\n",
    "\n",
    "\n",
    "X_corr = data\n",
    "y_corr = data_header['Sex'].loc[data.index]\n",
    "\n",
    "for c in tqdm(X_corr.columns):\n",
    "    corr, pvalue = pointbiserialr(X_corr[c], LabelEncoder().fit_transform(y_corr.values))\n",
    "    if np.abs(corr) > 0.5:\n",
    "        print(c)\n",
    "\n",
    "data.values[:, 0].shape, data_header['Sex'].loc[data.index].values.shape\n",
    "# data['Sex']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "numerical_cols = data.iloc[:1].select_dtypes(include=[np.number]).columns\n",
    "data = data.replace(0, 1e-6)\n",
    "data = np.log(data)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_by_cv(df, threshold):\n",
    "    cv = df.std() / df.mean()\n",
    "    # print(cv)\n",
    "    low_cv_cols = cv[cv < threshold].index\n",
    "\n",
    "    if len(low_cv_cols) > 0:\n",
    "        print(f\"{len(low_cv_cols)} features have coefficient of variation below {threshold} and will be removed.\")\n",
    "        df = df.drop(columns=low_cv_cols)\n",
    "    else:\n",
    "        print(\"No features found with coefficient of variation below the threshold.\")\n",
    "    print(f\"Current amount of features is {len(df.columns)}\")\n",
    "    return df\n",
    "\n",
    "cv_threshold = 0.7\n",
    "data = filter_by_cv(data, cv_threshold)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.quantile(q=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fdir_processed = Path(\"/home/ar3/Documents/PYTHON/RNASeqAnalysis/data/interim\")\n",
    "data = data.astype(np.float32)\n",
    "data['Sex'] = data_header['Sex']\n",
    "data.to_csv(fdir_processed/'geuvadis.preprocessed.XY.csv')\n",
    "# dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove transcripts located on sex chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from gtfparse import read_gtf\n",
    "\n",
    "gtf_data = read_gtf(fdir_raw/'gencode.v44.annotation.gtf')\n",
    "gtf_data = gtf_data.to_pandas()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_data = gtf_data[['seqname', 'transcript_id']]\n",
    "transcripts_x = gtf_data.loc[gtf_data['seqname'] == 'chrX', 'transcript_id']\n",
    "transcripts_y = gtf_data.loc[gtf_data['seqname'] == 'chrY', 'transcript_id']\n",
    "transcripts_x = transcripts_x.unique() \n",
    "transcripts_y = transcripts_y.unique()\n",
    "\n",
    "pd.Series(transcripts_x).to_csv(fdir_processed/\"transcripts_X.csv\")\n",
    "pd.Series(transcripts_y).to_csv(fdir_processed/\"transcripts_Y.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "transcripts_x = pd.read_csv(fdir_processed/\"transcripts_X.csv\", index_col=0).values.ravel().tolist()\n",
    "transcripts_y = pd.read_csv(fdir_processed/\"transcripts_Y.csv\", index_col=0).values.ravel().tolist()\n",
    "\n",
    "\n",
    "data_noX = data.drop(columns=data.columns.intersection(transcripts_x))\n",
    "data_noY = data.drop(columns=data.columns.intersection(transcripts_y))\n",
    "data_noXY = data_noY.drop(columns=data.columns.intersection(transcripts_x))\n",
    "\n",
    "# data.columns.intersection(transcripts_y)\n",
    "\n",
    "data_noX.to_csv(fdir_processed/'geuvadis.preprocessed.Y.csv')\n",
    "data_noY.to_csv(fdir_processed/'geuvadis.preprocessed.X.csv')\n",
    "data_noXY.to_csv(fdir_processed/'geuvadis.preprocessed.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LOAD PREPROCESSED DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fdir_processed = Path(\"/home/ar3/Documents/PYTHON/RNASeqAnalysis/data/interim\")\n",
    "\n",
    "# sex = \".\" + 'Y'\n",
    "sex = \"\"\n",
    "\n",
    "data = pd.read_csv(fdir_processed/f'geuvadis.preprocessed{sex}.csv', index_col=0)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pointbiserialr(X_corr[\"MSTRG.563.1\"], LabelEncoder().fit_transform(y_corr.values))\n",
    "\n",
    "# # sns.boxplot(data[[\"MSTRG.563.1\", 'Sex']], x='Sex', y=\"MSTRG.563.1\")\n",
    "# # X_corr[\"MSTRG.563.1\"].hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = data.drop(columns=['Sex'])\n",
    "y = data['Sex']\n",
    "\n",
    "test_size = 0.2\n",
    "random_state = 42\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(\n",
    "        X, y, test_size=test_size, random_state=random_state, shuffle=True)\n",
    "\n",
    "\n",
    "train_scaler = StandardScaler().fit(X_train)\n",
    "test_scaler = StandardScaler().fit(X_test)\n",
    "\n",
    "X_train = train_scaler.transform(X_train) \n",
    "X_test = test_scaler.transform(X_test)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import LabelEncoder\n",
    "label_encoder = LabelEncoder().fit(y_train)\n",
    "y_train = label_encoder.transform(y_train)\n",
    "# y_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train_, X_val, y_train_, y_val = train_test_split(X_train, y_train)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Train model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_threads = 6\n",
    "\n",
    "# CBC = CatBoostClassifier(loss_function='MultiClass',\n",
    "#                          od_pval=0.05,\n",
    "#                          thread_count=n_threads,\n",
    "#                          task_type=\"CPU\",\n",
    "#                          iterations=500,\n",
    "#                          learning_rate=0.03\n",
    "#                          #  devices='0'\n",
    "#                          )\n",
    "\n",
    "# CBC.fit(X_train_, y_train_, eval_set=(X_val, y_val), \n",
    "#         verbose=False, \n",
    "#         use_best_model=True, \n",
    "#         plot=True, \n",
    "#         early_stopping_rounds=20)\n",
    "\n",
    "# model = CBC\n",
    "\n",
    "model = XGBClassifier(early_stopping_rounds=20)\n",
    "model.fit(X_train, y_train, eval_set=[(X_val, y_val)])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ml_models_fdir = Path(\"/home/ar3/Documents/PYTHON/RNASeqAnalysis/models\")\n",
    "\n",
    "if not (ml_models_fdir/'catboost').is_dir():\n",
    "    (ml_models_fdir/'catboost').mkdir()\n",
    "\n",
    "saved_model_filename = f\"geuvadis_{sex}.cbm\"\n",
    "model.save_model(fname=ml_models_fdir/'catboost'/saved_model_filename, \n",
    "               format='cbm', export_parameters=None, pool=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sex = 'XY'\n",
    "saved_model_filename = f\"geuvadis_{sex}.cbm\"\n",
    "model.load_model(fname=ml_models_fdir/'catboost'/saved_model_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay,  RocCurveDisplay\n",
    "\n",
    "pred = model.predict(X_test)\n",
    "pred_prob = model.predict_proba(X_test)\n",
    "\n",
    "ConfusionMatrixDisplay(\n",
    "    confusion_matrix(label_encoder.transform(y_test),\n",
    "                 pred)\n",
    ").plot()\n",
    "\n",
    "RocCurveDisplay.from_predictions(\n",
    "    label_encoder.transform(y_test), pred_prob[:, 1]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_importance_df = pd.DataFrame({\n",
    "    'Feature': data.drop(columns='Sex').columns,\n",
    "    'Importance': model.feature_importances_\n",
    "    # 'Importance': XGB.feature_importances_\n",
    "})\n",
    "feature_importance_df = feature_importance_df.sort_values(\n",
    "    by='Importance', ascending=False)\n",
    "\n",
    "\n",
    "feature_importance_df.iloc[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import shap\n",
    "shap.initjs()\n",
    "\n",
    "explainer = shap.TreeExplainer(model)\n",
    "shap_values = explainer(pd.DataFrame(X_train_, columns=data.columns.drop('Sex')))\n",
    "\n",
    "# visualize the first prediction's explanation\n",
    "\n",
    "# shap.plots.force(explainer.expected_value, shap_values)\n",
    "\n",
    "# shap.dependence_plot(\"MSTRG.563.1\", shap_values.values, X_train_, interaction_index=\"HouseAge\")\n",
    "shap.plots.beeswarm(shap_values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['Sex'], label_encoder.transform(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
